### ISSUE TITLE
[Question] Reduce memory requirement in batch variational GPs with different evaluation times

### ISSUE THREAD

**Issue by flcello** [cornellius-gp/gpytorch#1765]
Hi,

I have two questions concerning the reduction of memory usage in variational GPs, specifically Cox processes:

**First question:** I'm considering a Cox process like in [this example](https://docs.gpytorch.ai/en/stable/examples/07_Pyro_Integration/Cox_Process_Example.html), running on a GPU. I noticed that calling the model does not use much memory, but calling the guide does. The memory usage of calling the guide grows with the number of arrivals or quadrature points passed to it. With approximately 30 000 arrivals my GPU memory is used up. I know that I could do subsampling, but is there another way to reduce memory consumption? I assume that the problem is not the data itself because it's just a shape [30 000] tensor that should easily fit in memory. I mean something like using a different variational distribution or so.

**Second (and more important) question:** Now I have different datasets and want to train one Cox process per dataset. However, this has to be done in a single Pyro model because I have additional latent variables that are global in the sense that they are the same for every dataset. To accomplish this I tried to use the `batch_shape` argument in the variational strategy, distribution and mean / covariance modules, as described in the [docs](https://docs.gpytorch.ai/en/stable/examples/04_Variational_and_Approximate_GPs/SVGP_Multitask_GP_Regression.html) (but without the `IndependentMultitaskVariationalStrategy` wrapper). The forward method then takes a tensor of evaluation times and samples from all batch GPs at these times. But in the Cox process example, having different datasets means that the different GPs also need to be evaluated at different arrival times. My solution so far was to simply concatenate the arrival times from all datasets, evaluate each batch GP at all times, and disentangle the values afterwards. However, I noticed that this drastically limits the number of datasets I can use: With my solution I have a total of `(batch_size * n_arrivals_all_datasets_concatenated)` GP evaluations, resulting in very high memory usage. It would be desirable to be able to evaluate the batch GPs at different arrival times, resulting in only `(n_arrivals_dataset_1 + n_arrivals_dataset_2 + ... + n_arrivals_batch_size)` evaluations and therefore also large memory savings. Is there a way to do this? If not, are you planning to implement something like this in the future?

I would be very happy to get this solved, because I enjoy using GPyTorch a lot!

**Comment by wjmaddox** [comment]
Without a minimal code example, I don't fully follow but I wonder if you could do two non-subsampling based approaches to reduce the overhead. Instead using a batch GP and predicting at all arrival times, I wonder if you could setup the kernel to be something like a product of the covariance between data points and "datasets" like this: `RBFKernel(active_dims=range(0, d-1)) * IndexKernel(num_tasks=datasets, active_dims=d-1)`.
This is the elementwise way of writing a multi-task kernel -- in the fully observed setting (all data points are observed for all tasks) it produces Kronecker structure.

Then, your test points wouldn't require a batch GP (and you could also detach the inter-task covariance plus enforce it to be diagonal).

**Comment by flcello** [comment]
Thanks! Could you explain what you mean with `d` in `active_dims`? I see that I could use a multi-task kernel if I had the same arrival times for every task. But I don't quite understand yet how to go from this to the case where I have different arrival times in every task.

**Comment by wjmaddox** [comment]
Sorry i missed this until now. 

By `d` I mean the dimensionality of your input data, in this setting it'd be at least two (you have `n_arrivals_all_datasets_concatenated x features` tensor where features is at least two -- time + batch identifier).

Sub-sampling by datasets could probably be accomplished via either a `BatchIndependentVariationalStrategy` or just a naive for loop across each dataset when training the model.

**Comment by flcello** [comment]
Never mind, thanks a lot for taking the time.

I have a code example now. It's really the same as [the Cox process tutorial](https://docs.gpytorch.ai/en/stable/examples/07_Pyro_Integration/Cox_Process_Example.html) except that it's extended to `M` independent processes for different datasets. The problem is that if I make `M` too large, I run out of memory. That's because all points (data and quadrature) are passed to every batch dimension when calling `pyro_guide()` or `pyro_model()`, the memory requirement grows linearly with `M`. So at the moment I think the best solution would be to be able to subsample the `M` batch dimensions. But for this I can't just subsample the data that I pass to `pyro_guide()` or `pyro_model()`, because still all batch dimensions would be evaluated at those points. What I probably need is a way to tell `pyro_guide()` or `pyro_model()` which batch dimensions I want to be evaluated. Do you know if that's possible?
```
class Independent_Cox_Processes(gpytorch.models.ApproximateGP):
    def __init__(
        self,
        X: torch.tensor,
        n_inducing: int=50,
        n_quadrature: int=100,
        device='cpu',
    ):
        """Independent Cox processes for multiple datasets.
        
        :param X: tensor of shape (N_1 + ... + N_M, 2) with arrival times (all
        datasets) in 0th column and corresponding dataset indices in 1st
        column. Arrival times should be rescaled to the unit interval.
        :param n_inducing: number of inducing points per process
        :param n_quadrature: number of quadrature points in likelihood integral
        :param device: 'cpu' or 'cuda:x'
        """
        self.n_inducing = n_inducing
        self.n_quadrature = n_quadrature
        self.device = device
        self.X = X.to(device=self.device)

        self.M = X[:, 1].unique().shape[0] # number of independent processes

        # number of arrivals per dataset (equals mean intensity)
        self.mean_intensity = torch.zeros(self.M, device=self.device)
        for m in range(self.M):
            self.mean_intensity[m] = X[X[:, 1] == m].shape[0]

        # evenly spaced quadrature points for Poisson process likelihood integral
        self.quadrature_points = torch.linspace(0, 1, n_quadrature, device=self.device)

        # initial inducing points
        inducing_points = torch.rand([self.M, n_inducing, 1], device=self.device)

        var_dist = gpytorch.variational.CholeskyVariationalDistribution(
            num_inducing_points=n_inducing,
            batch_shape=torch.Size([self.M]),
        )

        variational_strategy = gpytorch.variational.VariationalStrategy(
            model=self,
            inducing_points=inducing_points,
            variational_distribution=var_dist,
        )

        super().__init__(variational_strategy=variational_strategy)

        self.mean_module = gpytorch.means.ZeroMean(
            batch_shape=[self.M],
        )

        self.covar_module = gpytorch.kernels.RBFKernel(
            batch_shape=[self.M],
        )

    def forward(
        self,
        x: torch.tensor,
    ):
        mean = self.mean_module(x)
        covar = self.covar_module(x)

        return gpytorch.distributions.MultivariateNormal(mean, covar)

    def guide(
        self,
    ):
        # draw samples from q(f) at all arrival times and at quadrature times
        with pyro.plate('datasets_plate', dim=-2, device=self.device):
            with pyro.plate('times_plate', dim=-1, device=self.device):
                pyro.sample(
                    name='function_samples',
                    fn=self.pyro_guide(
                        torch.cat([self.X[:, 0], self.quadrature_points], -1)
                    )
                )

    def model(
        self,
    ):
        pyro.module('gp', self)

        # draw samples from p(f) at all arrival times and at quadrature times
        with pyro.plate('datasets_plate', dim=-2, device=self.device):
            with pyro.plate('times_plate', dim=-1, device=self.device):
                function_samples = pyro.sample(
                    name='function_samples',
                    fn=self.pyro_model(
                        torch.cat([self.X[:, 0], self.quadrature_points], -1))
                ).view(-1, self.M, self.X.shape[0] + self.n_quadrature)

        # make intensity values positive and scale by mean intensity per process 
        intensity_samples = function_samples.exp() * self.mean_intensity.view(1, self.M, 1)

        # divide arrival time intensities in processes to which they belong
        arrival_intensity_samples = dict()
        for m in range(self.M):
            inds = (self.X[:, 1] == m).nonzero()
            arrival_intensity_samples[m] = intensity_samples[:, m, inds].squeeze(-1)

        quadrature_intensity_samples = intensity_samples[:, :, self.X.shape[0]:]

        # log likelihood
        arrival_log_intensities = torch.zeros(function_samples.shape[0])
        for m in arrival_intensity_samples.keys():
            arrival_log_intensities += arrival_intensity_samples[m].log().sum(dim=-1)

        est_num_arrivals = quadrature_intensity_samples.mean(dim=-1).sum(dim=-1)

        log_likelihood = arrival_log_intensities - est_num_arrivals
        pyro.factor('log_likelihood', log_likelihood)
```

**Comment by gpleiss** [comment]
If the processes are independent, then you could train `M` different models in series (or `M/B` models with batch size `B`). This should get around your memory issues.

**Comment by flcello** [comment]
Yes, in the simplified model that I posted this would make more sense. But in the actual model I have global Pyro parameters that I want to learn with all datasets jointly. So separate models are not an option unfortunately.

**Comment by gpleiss** [comment]
This is being addressed (somewhat) by #1769, where you can supply the task indices associated with each input. This will return 1 output per input, rather than `num_tasks` outputs per input.

You could use a `LMCVariationalStrategy`, which will be more efficient. (You will have `num_latent < M` GPs, and then each of the M tasks will be a linear combination of these `num_latent` GPs. This should hopefully result in quite a bit of memory savings.

**Comment by flcello** [comment]
#1769 is definitely going in the right direction for me. Do you think something similar would be possible with the batch dimensions in the standard `VariationalStrategy`? Ideally I would want to have an argument `batch_inds` in `pyro_guide` to only evaluate the inputs on a subset of the batch dimensions. I played a bit with the code but I don't know if I understand it well enough at the moment.

**Comment by gpleiss** [comment]
I don't think it makes sense in the standard VariationalStrategy, but [IndependentMultitaskVariationalStrategy](https://docs.gpytorch.ai/en/latest/variational.html#independentmultitaskvariationalstrategy) may be exactly what you want.

**Comment by flcello** [comment]
Thanks, it looks like this solves my problem! But to use `task_indices` with the low-level Pyro interface the `pyro_guide` and `pyro_model` methods in `_PyroMixin` need `**kwargs`, right?
